File: leetcode/tree/94. Binary Tree Inorder Traversal/js/main.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */

// Time O(N)
// Space O(N)
const inorderTraversal = function (root) {
  let nodeStack = [];
  let node = root;
  const resArr = [];

  while (nodeStack.length > 0 || node !== null) {
    while (node !== null) {
      nodeStack.push(node);
      node = node.left;
    }

    node = nodeStack.pop();
    resArr.push(node.val);
    node = node.right;
  }
  return resArr;
};

File: leetcode/tree/94. Binary Tree Inorder Traversal/py/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Time O(N)
# Space O(N)
class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        st, ans = [], []
        node = root

        while st or node:
          while node:
            st.append(node)
            node = node.left

          node = st.pop()
          ans.append(node.val)
          node = node.right
          
        return ans